enhance docs/WAIVERS.md, describe the waiving algorithm (#208)
RHSecurityCompliance/contest: docs/WAIVERS.md

    rhel == 8 and 'some thing' in note and env('INFRA') == 'jenkins'
```

## Collecting a list of sections

Sections (as defined above) are gathered from multiple waiver files
in alphanumerical order, and - within each file - from the top.

This creates a one big unified sequential ordered list of sections
that is later used for waiving.

## How a result is waived

> When `results.report()` is called by test code to report a new result,
> the waiving logic looks at the status (`pass`, `fail`, etc.).

Reviewer: Don't mention test implementation details such as `results.report()`. Results reviewer might not be familiar with tests implementations and those lines basically say "see test implementation".

Use rather something like
```suggestion
For each individual result before reporting it, the waiving logic
looks at the status (`pass`, `fail`, etc.).
```
or
```suggestion
Before a result (e.g. rule result or overall test result) is reported,
the waiving logic looks at the status (`pass`, `fail`, etc.).
```

If the status is `info` or `warn`, no further processing is done, and
the result remains intact.  
All other statuses continue below.

The waiving logic then goes through the big list of sections (gathered
above), **from beginning to end**, and tries to match the test name
contained within the result, against all regexps in each section.

For example, a test name of `/some/result/name`, when evaluated against
```
/different/name.*
    True

> /some/unrelated/string
> /some/.+/name
>     True

Reviewer: Switch the regexes and use more interesting condition than just `True`

```suggestion
/some/.+/name
/some/unrelated/string
    rhel >= 8
```

It will be useful for further explanation

Author: Agree with the non-True condition, but I wanted to emphasize that it's not the first regex that has to match - it can be any of them. And for that, putting it second seemed better.

Reviewer: And I'd like to show it doesn't have to be right next to the regex :D What about putting 3 regexes with 2nd as the match?

Author: Sure.

```
would match the second section, and its `/some/.+/name` regexp.

**If no section matches on a regexp, no further processing is done,
and the result remains intact.**

> When a section is matched on regexp, its python expression is evaluated.
> If that expression returns `False`, the section is skipped, and further
> sections (in the big unified list of sections) are tried, to see if
> at least one of them could match.

Reviewer: Refer the example above here and add extra details
```suggestion
When a section is matched on regexp, its python expression is evaluated.
In our example, `/some/result/name` matched `/some/.+/name` regex. The
python expression to be evaluated for this section is `rhel >= 8`.
If that expression returns `False`, the section is skipped, and result
remains intact for now. Waiving continues with further sections (in the
big unified list of sections) to look for next matching regex.
```

**If no section matches on both a regexp AND python expression returning
True, no further processing is done, the result remains intact.**

> When the result matches on both regexp + python of one section, the result
> is subject to waiving, and **no further sections are evaluated for that
> result**, the processing stops on the first matching section.

Reviewer: ```suggestion
When the result matches on both regexp + python expression, the result
is waived, the processing stops, and waived result (e.g. `warn (waived fail)`)
reported to user. **No further sections are evaluated for that result.**
```

Author: I think I like my version better - `warn (waived fail)` doesn't appear anywhere (HTML, console, or this document, etc.) and I would really like to emphasize that "the processing stops on the first matching section", to make the reader realize the atomic unit of processing is a section.

Reviewer: You don't have to use exactly what I wrote, we can combine your version with mine. We don't have to mention `warn (waived failed)`. But to me, "the processing stops on the first matching section" says the same as "no further sections are evaluated for that result".

Author: I rewrote it slightly, and removed that redundant addition.

> If a result got through all that, the waiving logic decides based on the

Reviewer 2: ```suggestion
If a result got through all that and a matching section was found, the waiving logic decides based on the
```

> If a result got through all that, the waiving logic decides based on the
> reported status (`pass`, `fail`, etc.):

Reviewer: Too complicated wording, I'd simplify it

```suggestion
If a result got matched by a waiver, the waiving logic decides on reported status:
```

- `fail` is changed to `warn`, and "waived fail" is added to `note`
- `error` is changed to `warn`, and "waived error" is added to `note`

> Further, if a waiver returns `strict=True` via `Match()`, or if
> `CONTEST_STRICT_WAIVERS` is set:

Reviewer: Add extra info about permissive(default) / strict

```suggestion
By default, waivers are permissive causing `pass` result is
reported as `pass (waived pass)` result.

That might be changed to strict via waiver `strict=True`, e.g.
`Match(rhel >= 8, strict=True)`, or running Contest with
`CONTEST_STRICT_WAIVERS`. Adds waiving logic for `pass`:
```

Author: Again, I think I like my version better - it doesn't try to make `pass` a special free-form text thing. I'm not sure, "Adds waiving logic for `pass`"  is clearer here.

And `pass (waived pass)` also doesn't appear anywhere.

But I agree that `Match(rhel >= 8, strict=True)` should probably be mentioned, since I removed the section explaining how to use `strict=True`.

Reviewer: Sure, you don't have to use exactly what I wrote. But I'd like to have some parts at least clearer.
For example:
```
By default, waivers are permissive causing if `pass` result matches a waiver, the result is intact.
```
or
```
By default, waivers are permissive causing if `pass` result matches a waiver, no `fail`/`error` is reported.
```
"Adds waiving logic for `pass`" can be also changed to something better like "Changes `pass` waiving behavior".
